Default exception handler

anacierdem opened the pull request:
~~Fix console new line bug~~

~~- Previously it was adding an extra space when the current line is exactly CONSOLE_WIDTH~~

Interrupt handler improvements

    - Remove some unnecessary nops
    - No need to disable interrupts while in the interrupt handler
    - No need to save/restore k0 & k1
    - Do not restore fcr31 register, it should be always cleared by the ex handler.
    - Store FCR31 as a 32bit value

Add cop0 & cop1 macros and start using them

Remove inline wait functions from the header

Implement and register a default exception handler

I'll be implementing a test for the interrupt handler next.

anacierdem committed:
Fix console new line bug

- Previosly it was adding an extra space when the current line is exactly CONSOLE_WIDTH

anacierdem committed:
Interrupt handler improvements

- Remove some unnecessary nops
- No need to disable interrupts while in the interrupt handler
- No need to save/restore k0 & k1
- Do not restore fcr31 register, it should be always cleared by the ex handler.
- Store FCR31 as a 32bit value

anacierdem committed:
Add cop0 & cop1 macros

anacierdem committed:
Remove inline wait functions from the header

anacierdem committed:
Start using new macros

anacierdem committed:
Implement and register a default exception handler

rasky commented on include/cop0.h:
```diff
@@ -50,19 +50,49 @@
+#define C0_INTERRUPT_4      0x00001000
+#define C0_INTERRUPT_5      0x00002000
+#define C0_INTERRUPT_6      0x00004000
+#define C0_INTERRUPT_7      0x00008000 // Timer
```
Maybe we should call them `C0_INTERRUPT_RCP` and `C0_INTERRUPT_TIMER`.

rasky commented on src/exception.c:
```diff
@@ -36,6 +40,145 @@ void register_exception_handler( void (*cb)(exception_t*))
+	// will gracefully restore interrupt state so that we can
+	// continue using interrupt dependent functionalities
+	// such as console updates.
+	C0_WRITE_EPC(&exception_halt);
```
Did you try using `&abort` here? So that we don't need a specific function. But if we need to actually call `console_render()`, then a function is required.

anacierdem replied:
Ill try that but I suspect it wont work

anacierdem replied:
Unfortunately it does not work. I tried to comment on why it does not work without the interrupts. It may not be 100% accurate and I think it is not worth pursuing further just to remove the extra function. We may have a look again when we update he display system someday :)

rasky commented on src/inthandler.S:
```diff
@@ -17,17 +17,13 @@ inthandler:
 	sw k1,($1)
-	la $1, ~1
-	and k1,$1
-	mtc0 k1,C0_SR
```
I think it made sense to turn off the interrupt bit here. Doesn't it?

anacierdem replied:
Why do you think so? EXL is already set at this point and I dont think an explicit disable is necessary.

rasky replied:
Ok I think you're right, it's probably not required.

rasky commented on src/inthandler.S:
```diff
@@ -98,8 +92,7 @@ inthandler:
 	sdc1 $f24,saveFR24
 	sdc1 $f25,saveFR25
-	sdc1 $f26,saveFR26
-	sdc1 $f27,saveFR27
```
`$f26` and `$f27` I guess we need to save?

anacierdem replied:
Oops..

anacierdem committed:
Fix warnings and printing issues

rasky commented on src/inthandler.S:
```diff
@@ -229,8 +221,6 @@ endint:
 	nop
 
-	ctc1 $30,$f31
-
```
Uhm not sure about this. I think that, in general, we need restore `$fc31` in case the interrupt code modifies it. For instance, it contains floating point flags. I see that for the specific case of FPU exception, you need to mask some flags but maybe that should be handled specially, while still allowing the other bits to be restored.

anacierdem replied:
Maybe we should set bits except the ones that needs to be reset by the handler. That would be better youre right.

anacierdem commented on src/exception.c:
```diff
@@ -36,6 +40,145 @@ void register_exception_handler( void (*cb)(exception_t*))
+
+			fprintf(stdout, "Floating-point registers\n");
+			for (int i = 0; i<32; i++) {
+				fprintf(stdout, "%02u:%08llX ", i, ex->regs->fpr[i]);
```
We need 32 bit values here if we dont switch to a higher res.

anacierdem replied:
Which wont make any sense for doubles.

anacierdem commented on src/exception.c:
```diff
@@ -36,6 +40,145 @@ void register_exception_handler( void (*cb)(exception_t*))
+			fprintf(stdout, "BadVAddr: %08lX\n", C0_READ_BADVADDR());
+		break;
+		case EXCEPTION_CODE_COPROCESSOR_UNUSABLE:
+			fprintf(stdout, "COP:%1lu\n", (cr >> 28) & 3);
```
Use macros here as well

anacierdem commented on src/exception.c:
```diff
@@ -36,6 +40,145 @@ void register_exception_handler( void (*cb)(exception_t*))
+		sr & C0_INTERRUPT_7 ? 1 : 0);
+
+	switch(ex_code) {
+		case EXCEPTION_CODE_STORE_ADDRESS_ERROR:
```
Also combine with other addressing errors

anacierdem reviewed (commented):
Self review

anacierdem committed:
Fixes and updates

anacierdem committed:
Print most of the useful information to the screen
- Increase console resolution and size
- Update console paddings to match new size better
- Previous console fix was bogus, revert it
- Save mistakenly removed FP registers
- Restore FCR31 bits except cause

anacierdem committed:
Rename interrupt macros

anacierdem committed:
Fix incorrect store/load for FCR31

rasky commented on include/console.h:
```diff
@@ -41,9 +41,9 @@
+#define CONSOLE_WIDTH       74
 /** @brief The console height in characters */
-#define CONSOLE_HEIGHT      25
+#define CONSOLE_HEIGHT      30
```
Self note: I want to check this on my CRT. I think borders in general help because of the typical overscan of CRTs. I'm not sure we should remove them

anacierdem replied:
Vertical padding does not seem to be an issue on the highest res mode. We even increased the horizontal padding as it was not enough for 640x240. Please test on your screen and let's see. We are already on the limits of available screen space so I hope it looks good :)

rasky commented on src/console.c:
```diff
@@ -268,7 +268,7 @@ static void __console_render()
             /* Draw to the screen using the forecolor and backcolor set in the graphics
              * subsystem */
-            graphics_draw_character( dc, 20 + 8 * x, 16 + 8 * y, t_buf );
+            graphics_draw_character( dc, 24 + 8 * x, 8 * y, t_buf );
```
Maybe expose the margins as defines as well, so that we can reason upon them better?

rasky commented on src/exception.c:
```diff
@@ -36,6 +42,159 @@ void register_exception_handler( void (*cb)(exception_t*))
+					C1_CAUSE_INVALID_OP |
+					C1_CAUSE_NOT_IMPLEMENTED
+				)
+			);
```
So to clarify: this is only necessary because we need the hack of returning from the exception to call console_render, right? In principle if we fix that, we could remove these lines, right? If so, I'll add a comment for later when we get around and fix the console.

anacierdem replied:
Exactly, we can remove them from the default handler if we fix that.

rasky commented on src/exception.c:
```diff
@@ -36,6 +42,159 @@ void register_exception_handler( void (*cb)(exception_t*))
+	fprintf(stdout, "cr:%08lX (COP:%1lu BD:%lu)\n", cr,  C0_GET_CAUSE_CE(cr), cr & C0_CAUSE_BD);
+	fprintf(stdout, "sr:%08lX FCR31:%08X BVAdr:%08lX \n", sr, (unsigned int)fcr31, C0_READ_BADVADDR());
+	fprintf(stdout, "--------------------------------------------------------------------------");
+	fprintf(stdout, "FPU     IOP UND OVE DV0 INV  NI  | INTs    sw0 sw1 ex0 ex1 ex2 ex3 ex4  tm");
```
Maybe the last string can be "tmr"

rasky commented on src/interrupt.c:
```diff
@@ -594,9 +594,9 @@ void init_interrupts()
 
         /* Enable interrupts systemwide. We set the global interrupt enable,
-           and then specifically enable RCP interrupts (IM2). */
+           and then specifically enable RCP interrupts (INTERRUPT_2). */
```
INTERRUPT_2 in the command doesn't refer anymore to a macro, right? Maybe just remove it as the macro now has RCP in the name

anacierdem replied:
HMm I missed this one apparently

rasky commented on src/inthandler.S:
```diff
@@ -69,8 +64,7 @@ inthandler:
 	cfc1 $30,$f31
-	nop
-	sd $30,saveFC31
+	sw $30,saveFC31
```
I assume you have tested this specific NOP removal on real hardware right?

anacierdem replied:
True, it is still working on real hardware

rasky commented on src/inthandler.S:
```diff
@@ -226,9 +219,18 @@ endint:
+	cfc1 k0, $f31
+	and k0, 0x3F000
+	or k0, ~0x3F000
+	and $30, k0
```
OK this makes sense if we keep the current approach where the exception handler registered by the application has the task to also clear the exception (contrary to interrupts, where the interrupt is acknowledge by libdragon itself before calling the handler).

So I would maybe document at the top `exception.c` and in the comment of `register_exception_handler` that the handler must clear the exception in case it's required to avoid re-triggering.

anacierdem replied:
I think it is better to leave this to the consumer indeed. Floating point exceptions are disabled by default and user may never decide to enable them.

I initially had an alternative implementation where I removed the const specifier from the exception struct so that the ex handler can change the contents to be loaded back. I started to believe that is more useful now.

If we do that, we can remove this. We will only give write access to epc and fcr31 (and maybe sr and cr) so that the custom handler won't have to use inline assembly and at the same time it means we completely inverted the control to the consumer.

What do you think @rasky?

rasky replied:
I think that in general exception handlers might need to modify registers. It’s a pretty advanced use case, but why not. So maybe I’d simply drop const everywhere and document that it can be done.

rasky reviewed (approved):
Great work! I left some minor comment. 

The only thing that really concerns me is using the full resolution on a CRT for the console by default. I would be less concerned if you disabled the margin just for the exception handler, but I would expect the console by default to be more usable with some margin.

anacierdem commented:
Unfortunately current console subsytem is not easy to extend. Will need a re-write to make it more customizable. I'm not a fan of increasing the default resolution. Maybe we should do it with explicit text render calls.

anacierdem commented:
I have also strting adding tests. Lets keep them in a single PR. It will take some time but will be worth it IMO.

anacierdem committed:
Add basic ex handling test

- Added C0_READ_EPC cop0 macro

anacierdem committed:
Fix $f0 & add hi & lo

anacierdem committed:
Ex handler should manipulate registers via the exception struct

- Removed const specifier from the regs member
- Disable benchmarking for test_exception
- Don't save/restore unnecessary registers
- Add description for register_exception_handler
- Do proper cleanup

anacierdem committed:
Update console paddings

- Convert paddings into macros
- Update default handler to use the exception struct
- Update exception handler to fit console
- Save gp for display purposes
- Re-enable interrupt disabling instruction

anacierdem commented:
Added tests & improved console paddings. The only remaining issue is the increased timer benchmark time, for which I couldnt find any lead yet.

rasky commented on src/exception.c:
```diff
@@ -21,13 +27,26 @@
+ * a re-trigger of the same exception and updating the EPC. To manipulate the
+ * registers, update the values in the exception_t struct. They will be
+ * restored to appropriate locations when returning from the handler. Setting
+ * them directly will not work as expected as thhey will get overwritten.
```
```suggestion
 * them directly will not work as expected as they will get overwritten.
```

rasky commented on src/exception.c:
```diff
@@ -21,13 +27,26 @@
+ * prevent re-triggering the exception they should be cleared by the handler.
+ * k0 ($26), k1 ($27), s0-s7 ($16-$23), and gp ($28) are not saved/restored
+ * and will not be available in the handler. These are removed to gain some
+ * performance as they are already saved by GCC when necessary. Theoretically
```
I think that for *interrupts* it makes sense not to save those registers as they are callee-saved (BTW, the same I think it applies to f15-f31 but should be verified). Instead for exceptions, one might want to have a look at the whole state of all regs. So maybe the interrupt handlers should save/restore those only after it has verified that it's an exception.

Anyway, this can be improved in the future, no need to do it now.

anacierdem replied:
https://github.com/DragonMinded/libdragon/issues/129

rasky reviewed (approved):
Great work!

anacierdem committed:
Update macros

- Remove  C0_WRITE_EPC
- Remove duplicate WRITE_TICKS from tests

anacierdem committed:
Guard against FCR31 flag changes

anacierdem committed:
Do not restore cause register through the struct

- It was restoring the pending bits after they were handler by the library
- Added macros to read/write CR in case it is needed.

anacierdem committed:
Minor fixes

anacierdem committed:
Fix compilation issue

anacierdem committed:
Stop skipping register saves

- Add s0-s7 back to the ex report
- Update the tests to match the new configuration
- Update comments to match

anacierdem committed:
Update testrom to match new console better

anacierdem commented:
That benchmark works fine with this state. I didn't find the cause or applied a solution though. So we better keep an eye on that. I'm mostly happy with this PR. Can merge when @rasky approves :)

rasky commented:
Fine by me, I tested it everywhere and it works. I previously told you that I had an unexpected FPU exception but it turned out that was really a bug in my code so this PR actually helped me debugging a real issue.
